src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -327,6 +352,36 @@ protected Object invokeStep(StepDescriptor d, String name, Object args) {
         }
     }
 
+    private void logInterpolationWarnings(Set<String> interpolatedStrings, CpsStepContext context, TaskListener listener) throws IOException, InterruptedException {
+        if (interpolatedStrings.isEmpty()) {
+            return;
+        }
+        EnvVars contextEnvVars = context.get(EnvVars.class);
+        EnvironmentExpander contextExpander = context.get(EnvironmentExpander.class);
+        if (contextEnvVars == null || contextExpander == null) {
+            return;
+        }
+
+        List<String> scanResults = contextExpander.getSensitiveVariables().stream()
+                .filter(e -> interpolatedStrings.stream().anyMatch(g -> g.contains(contextEnvVars.get(e))))
+                .collect(Collectors.toList());
+
+        if (scanResults != null && !scanResults.isEmpty()) {
+            listener.getLogger().println("The following Groovy string(s) may be insecure. Use single quotes to prevent leaking secrets via Groovy interpolation. Affected variable(s): "  + scanResults.toString());
```

Comment:
This message is a little confusing since we don't actually report the strings with bad interpolation themselves, only the environment variables that were used. We might be able to replace the sensitive content with the name of the variable (or maybe reuse `ArgumentsAction` somehow?), and I think it would help as well to display the name of the step that was being invoked so we can give some context. I think the ideal message would be something like this:

> Warning: A secret was passed to \<STEP_NAME\> using Groovy String interpolation, which is insecure. Affected argument(s): "blah ${PASSWORD} blah", "foo ${PASSWORD_2} foo". See \<LINK\> for details."

If we had that information, I would store it all in `InterpolatedSecretsAction` as well (e.g. have the field be a `List<(String stepName, List<String unsafe argument (with secrets replaced with ${variableName})> )>)` so we can enrich the summary page to have the same information.

Reply:
So digging into `ArgumentsActionImpl`. It could be possible to use it to get specifically which arguments failed, but `getArguments()` displays all the step arguments as a hashmap and the level of nesting can vary. I could make changes to `ArgumentsActionImpl` where you can record which arguments were affected, but the sanitization process is deeply recursive that I am inclined to not do it.

Reply:
Using `ArgumentsActionImpl` to report warnings in 6a39c5c49ec80a658dbc9f0956d6c1c20c68f7fc